### PR TITLE
rustc: add QA documentation

### PR DESCRIPTION
**Pull request by Mic92**
###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @

### CODE REVIEW

**Comment by Mic92** [comment]
cc @andir @edolstra

File: pkgs/development/compilers/rust/1_39_0.nix
Comment (edolstra) [inline pkgs/development/compilers/rust/1_39_0.nix]: Don't really agree with this. For example, I have no ability to test on aarch64 or macOS.

File: pkgs/development/compilers/rust/1_39_0.nix
Comment (edolstra) [inline pkgs/development/compilers/rust/1_39_0.nix]: What does `fd` mean? And to what directory does `-f .` refer?

File: pkgs/development/compilers/rust/1_39_0.nix
Comment (edolstra) [inline pkgs/development/compilers/rust/1_39_0.nix]: What are the implications of this? E.g. if firefox fails to build, should we simply refrain from updating Rust?

Comment (Mic92) [reply pkgs/development/compilers/rust/1_39_0.nix]: Yes. Because this is a channel blocker. A way to fix it in that case is to to pin rust to an older version for firefox. However breaking firefox in unstable is undesired.

File: pkgs/development/compilers/rust/1_39_0.nix
Comment (Mic92) [reply pkgs/development/compilers/rust/1_39_0.nix]: @GrahamcOfBorg usually can build this in time. For aarch64 we have a beefy community builder: https://github.com/nix-community/aarch64-build-box
which I have access to and usually try to build upgrades.
For macOS alternatively I have a bunch of nixpkgs maintainer that I can ping instead. We followed this pattern for a while until you started merging rust upgrades without testing.

File: pkgs/development/compilers/rust/1_39_0.nix
Comment (Mic92) [reply pkgs/development/compilers/rust/1_39_0.nix]: `fd` is just a simple rust example that should build on all platforms.

**Comment by Mic92** [comment]
Other usual suspects in rust land: cc @symphorien @jD91mZM2 @zimbatm

**Comment by worldofpeace** [comment]
Is there any particular reason why rustc updates have been merged to master?

**Comment by Mic92** [comment]
Yeah. Maybe I should add this as well, we usually prefer to merge rustc updates to staging.

**Comment by worldofpeace** [comment]
> Yeah. Maybe I should add this as well, we usually prefer to merge rustc updates to staging.

Thanks, we should be doing this always. I can recall at least two occasions where this has taken up my time because there were urgent failures on master because of a rustc update. When it's on staging it isn't as urgent and we can fix it during its workflow.

**Comment by Mic92** [comment]
It now mentions, that rust updates should go to staging first and that one can ask other people for help to test rust on other platforms. In my experience this testing takes less than a week depending on how fast people respond but it is worth it to avoid unnecessary breakages in master.

**Comment by worldofpeace** [comment]
> It now mentions, that rust updates should go to staging first and that one can ask other people for help to test rust on other platforms. In my experience this testing takes less than a week depending on how fast people respond but it is worth it to avoid unnecessary breakages in master.

Oh, I know this my be rather nitpicky, but maybe just having a `README.md` or `QA-Instructions.md` in the dir would be better?

**Comment by jD91mZM2** [comment]
I think it makes sense to have it inline, as you can't possibly avoid opening that file when updating rust. I'm normally somewhat of an expert at failing to read documentation ;)

**Comment by LnL7** [comment]
Until recently we've only kept a single version of rust around. With 2 or more versions we could keep the default on a version compatible with firefox as well as make introducing versions much less intrusive.

**Review by worldofpeace — APPROVED** [review]

**Pull request merged**
